### ISSUE TITLE
[HOLD for payment 2024-10-07]  Workspace - Invite message comes back after deleting it

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#49899]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.41-2
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** applausetester+ds@applause.expensifail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Login to an account
2. Create a workspace
3. Go to the created workspace > Members
4. Click on invite member > Choose a user
5. On "add message" page try to delete the message

## Expected Result:
Message is deleted

## Actual Result:
The deleted message comes back

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/user-attachments/assets/b04cf507-1cde-4a47-a349-9c0f23c369c2

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by Shahidullah-Muffakir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
 when a user deletes the invite message in the workspace invite process, the message reappears unexpectedly. 
### What is the root cause of that problem?
The root cause of this problem is that the component was automatically resetting the welcome message to its default value, even when the user had intentionally deleted it
### What changes do you think we should make in order to solve the problem?
 we should update the  useEffect to add a check that skips resetting the invite message if it is already empty. This ensures that when the user deletes the message, it doesn't get reset to the default value. 

Add` !welcomeNote `check here: 
https://github.com/Expensify/App/blob/3047c1baa21f47825f5c1ffd652184341dd22256/src/pages/workspace/WorkspaceInviteMessagePage.tsx#L89-L92

to:
```
 useEffect(() => {
        if (isEmptyObject(invitedEmailsToAccountIDsDraft) || !welcomeNote) {
            return;
        }
```


https://github.com/user-attachments/assets/c1bf77cf-634a-4405-944b-a4aefe147855

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Invite message comes back after deleting it

### What is the root cause of that problem?
Since we set `getDefaultWelcomeNote` as a dependency of this hook, the hook is triggered every time the input's `onChange` event occurs

https://github.com/Expensify/App/blob/36346702f2e2e1621bb2b31d6a7d35cbe1c1d105/src/pages/workspace/WorkspaceInviteMessagePage.tsx#L94

### What changes do you think we should make in order to solve the problem?
We should remove `getDefaultWelcomeNote` from the dependency array to avoid re-renders when the input changes.
```diff
// .src/pages/workspace/WorkspaceInviteMessagePage.tsx#L94
    useEffect(() => {
        if (isEmptyObject(invitedEmailsToAccountIDsDraft)) {
            return;
        }
        setWelcomeNote(getDefaultWelcomeNote());
-    }, [getDefaultWelcomeNote, invitedEmailsToAccountIDsDraft]);
+    }, [invitedEmailsToAccountIDsDraft]);
```
and We need to validate this input, if necessary, to avoid empty values ("")
```diff
// .src/pages/workspace/WorkspaceInviteMessagePage.tsx#L119
        if (isEmptyObject(invitedEmailsToAccountIDsDraft)) {
            errorFields.welcomeMessage = translate('workspace.inviteMessage.inviteNoMembersError');
-        }
+        } else if (isEmptyObject(welcomeNote)) {
+            errorFields.welcomeMessage = 'Please enter a welcome note.';
        }
```

<summary>
<details>POC

https://github.com/user-attachments/assets/60e6acda-8cf1-4687-97e2-462e2df34f8b

</details>
</summary>

**Comment by Gonals** [comment]
Seems to be a frontend issue, so removing Web-E blocker

**Comment by dangrous** [comment]
Yeah this was caused by PR https://github.com/Expensify/App/pull/48660 and issue https://github.com/Expensify/App/issues/34929 . We can just remove that new `useEffect` but I'm not sure what the point of it was originally, so pinged the PR authors. Thanks!

**Comment by rojiphil** [comment]
> Yeah this was caused by PR #48660 and issue #34929 . 

@dangrous Yeah. Fixing an eslint error (i.e. using useOnyx instead of withOnyx) during some last-minute changes caused this issue. I think we can consider this as a separate issue. Please assign me here as I have the context.

> We can just remove that new `useEffect` but I'm not sure what the point of it was originally, so pinged the PR authors. Thanks!

We need the second `useEffect` to ensure `workspaceInviteMessageDraft` is populated from the Onyx. However, an edge case got left out i.e. when the user intentionally deletes the entire welcome note in which case we should not reset.
@Shahidullah-Muffakir [proposal](https://github.com/Expensify/App/issues/49899#issuecomment-2381620224) LGTM

**Comment by dangrous** [comment]
Yep checking for an empty welcome note makes sense to me. Can you raise that PR quick?

**Comment by rojiphil** [comment]
> Yep checking for an empty welcome note makes sense to me. Can you raise that PR quick?

Sure. Working on the PR now. Will update in about an hour

**Comment by jasperhuangg** [comment]
@rojiphil Can you please address that in a followup?

**Comment by rojiphil** [comment]
> I believe we still need to clear the message when the RHP closes. If we edit the message, close the RHP, and select members again, the edited message is shown instead of the default message.

@Krishna2323 The way it was working before was that if the welcome message text was cleared and we navigated away and return to the same page, the default welcome message was displayed. I tested this against a recent [ad-hoc build](https://49452.pr-testing.expensify.com/) generated before the changes were introduced. And the test video is demonstrated below. 

So, I think this is expected and there is no regression here. Setting empty text seems to be the way to get back to the default welcome message. 

cc @jasperhuangg 

https://github.com/user-attachments/assets/78760cb5-27db-4809-80fb-bf626200e506

**Comment by dangrous** [comment]
I think we might want to discuss this in Slack? I'm not sure if we have an expected behavior called out for this scenario. Personally, I would probably say:

* As long as the RHP stays open, the message should not change on its own (stay edited as edited, or empty as empty, or default as default). This would be true even if you're changing who you're inviting.
* Once the RHP is closed, it should reset to the default value.\

Again, this is just my opinion, but is probably what I'd expect? We can ask for some other opinions though.

**Comment by rojiphil** [comment]
@dangrous  Yeah. We can take some other’s opinion too. Meanwhile, I am just a little confused about the following:

 >	•	Once the RHP is closed, it should reset to the default value.\

Did you mean removing the `draft` feature so that it resets to `default` value always? The reason why I ask is that the `draft` feature involves saving the `draft` message for any further member invites done locally. This was implemented [here](https://github.com/Expensify/App/pull/31770). 

Or did you mean resetting to `default` value only when it is `empty`? The current behavior though is like this.

**Comment by JmillsExpensify** [comment]
Payment summary: $250 to @rojiphil for PR resolving this issue.

**Comment by Shahidullah-Muffakir** [comment]
@JmillsExpensify 
I wanted to kindly ask if I might also be eligible for compensation, as the solution I proposed in [My proposal](https://github.com/Expensify/App/issues/49899#issuecomment-2381620224) was implemented in the [PR](https://github.com/Expensify/App/pull/49955). Even @rojiphil acknowledged and approved my proposal in [here](https://github.com/Expensify/App/issues/49899#issuecomment-2383732941). Thank you.

**Comment by dangrous** [comment]
Oh got it! I wasn't aware of that specific behavior - so I take back what I said haha. I guess then I think it should just always stay as edited, no matter what.

But that does bring up the question of how to return to the default message, since that seems like a feature we might want.

Maybe:

* As long as the RHP stays open, the message should not change on its own (stay edited as edited, or empty as empty, or default as default). This would be true even if you're changing who you're inviting.
* Once the RHP is closed **and the field is empty**, it should reset to the default value.

What do we think? Alternate (more involved) solution would be adding a `Reset to default` or similar button

**Comment by dangrous** [comment]
bump on [this](https://github.com/Expensify/App/issues/49899#issuecomment-2398288344) - do we think that's the right way forward?

**Comment by rojiphil** [comment]
> I wanted to kindly ask if I might also be eligible for compensation, as the solution I proposed in [My proposal](https://github.com/Expensify/App/issues/49899#issuecomment-2381620224) was implemented in the [PR](https://github.com/Expensify/App/pull/49955). Even @rojiphil acknowledged and approved my proposal in [here](https://github.com/Expensify/App/issues/49899#issuecomment-2383732941). Thank you.

It’s correct that the mentioned proposal was implemented here to fix the issue. I would be happy to split the compensation and share a part with @Shahidullah-Muffakir.

**Comment by rojiphil** [comment]
> Maybe:
> 
> * As long as the RHP stays open, the message should not change on its own (stay edited as edited, or empty as empty, or default as default). This would be true even if you're changing who you're inviting.
> * Once the RHP is closed **and the field is empty**, it should reset to the default value.
> 
> What do we think? Alternate (more involved) solution would be adding a `Reset to default` or similar button

So, that brings before us the following options:
1. As long as the RHP stays open, keep the message as `empty when it is `empty`. Reset to `default` value only once the RHP is closed and the message is `empty`.
2. When set to `empty`, always keep the message as `empty`. And use a `Reset to default` or similar button to reset to default value. 
3. Do nothing. i.e. Reset to default value whenever the invite message page is shown and the message is empty. This is irrespective of whether the RHP stays open or closed.

I personally like Option 2 but let’s see what’s the consensus would look like here 
cc @JmillsExpensify @Expensify/design

**Comment by dangrous** [comment]
bumped design in Slack to see if we can figure something out!

**Comment by shawnborton** [comment]
Can someone share a video of the problem at hand or maybe a more detailed description of the full problem that needs to be resolved?

**Comment by dannymcclain** [comment]
Yeah a video would be super helpful. It also might be worth looking into how Status works—feels like it might be kinda similar.

**Comment by rojiphil** [comment]
> Can someone share a video of the problem at hand or maybe a more detailed description of the full problem that needs to be resolved?

Let me try to explain with a video and see if it helps.

The test video attached here demonstrates the current behaviour i.e. `option 3`. Here, after clearing the default message, when we navigate back to `Invite new members` view and again navigate to `Add message` view, it again displays the `default` message.  

Now, the thought is that since the user has set the message as `empty`, it should remain empty. The approach in `option 1` would be to keep the message `empty` itself until the RHP is closed (i.e. user closes the `Invite new members` page). If RHP is closed and we, again, navigate to `Add message`, we can show the `default` message.  

`Option 2` proposes to keep the message edited as `empty` always irrespective of whether the RHP is open or closed. And then rely on `Reset to default` or similar button to reset to default value.

https://github.com/user-attachments/assets/a847c2c9-d034-48ef-b59f-6fac2bf299fb

**Comment by shawnborton** [comment]
Hmm I don't feel too strongly here. I can see where it makes sense that if you get to the "Add message" screen and make edits, those edits should remain regardless if you navigate back to the invite page or not. But I would say we should only save a new default message once the actual invite gets sent.

**Comment by dannymcclain** [comment]
>But I would say we should only save a new default message once the actual invite gets sent.

I agree with this. The behavior you're showing in the video doesn't seem out of place to me. I tested it in the product and if you do actually write something in the field, we persist that edit if you navigate back to the participant selector and then move forward again. So generally that feels pretty right to me.

The only thing that feels weird is if I customize the message and then completely abandon the flow (close the RHP without sending an invite) the custom message persists if I go through the invite flow again from scratch. That feels a little weird to me that we're saving a custom message that never got sent.

**Comment by shawnborton** [comment]
> That feels a little weird to me that we're saving a custom message that never got sent.

Yeah, I definitely agree with that part.

**Comment by dangrous** [comment]
Okay so it sounds like we might be going for an option 1.5:

1.5. As long as the RHP stays open, keep the message as empty when it is empty. Reset to default value only once the RHP is closed **without an invite being sent** ~and the message is empty~.

Am I interpreting that correctly?

**Comment by dannymcclain** [comment]
Yes I think that makes the most sense to me.

**Comment by dubielzyk-expensify** [comment]
Catching up on it all and yeah that feels right to me too

**Comment by dangrous** [comment]
Awesome, thank you! @rojiphil, are you up for tackling that adjustment, or should we open it up to contributors again?

**Comment by rojiphil** [comment]
> But I would say we should only save a new default message once the actual invite gets sent.

> 1.5. As long as the RHP stays open, keep the message as empty when it is empty. Reset to default value only once the RHP is closed **without an invite being sent** ~and the message is empty~.

Hmm.. I am slightly confused here. Are we saying the following? Or am I missing something here?

As long as the RHP stays open, keep the messages as `edited` (even if it is set to `empty`). And save the `edited` message as `invite message draft` only on an actual invite. Once the RHP closes without an actual invitation, discard the `edited` message. And when the RHP is shown again, show either the previously saved `invite message draft` or show the `default message` if no custom message is set.

**Comment by dannymcclain** [comment]
I think you've outlined it exactly how I was understanding it. 👍

**Comment by rojiphil** [comment]
> are you up for tackling that adjustment, or should we open it up to contributors again? 

@dangrous I think it’s best to open it up to contributors again to get the best proposal as this does not seem like a minor adjustment.

**Comment by dangrous** [comment]
Okay cool! I may actually just make a new issue for it, so we can close this one out / pay. It's more of a clarification on expected behavior than a regression or anything that should be handled here.

I'm also going to hold it on https://github.com/Expensify/App/issues/49996 which is related and should make things easier.

**Comment by dangrous** [comment]
Made https://github.com/Expensify/App/issues/51096!

**Comment by rojiphil** [comment]
> Okay cool! I may actually just make a new issue for it, so we can close this one out / pay.

Yeah. Since another issue is created for the enhancement, I think we can pay as per [this comment ](https://github.com/Expensify/App/issues/49899#issuecomment-2408426689) and close this one out. cc @JmillsExpensify

**Comment by JmillsExpensify** [comment]
Ok, so based on that comment, we'd have:

 - Contributor: $125 @rojiphil via Upwork
 - Contributor: $125 @Shahidullah-Muffakir via Upwork

**Comment by JmillsExpensify** [comment]
Offers sent to both via Upwork.

**Comment by rojiphil** [comment]
> Offers sent to both via Upwork.

Accepted the offer. Thanks

**Comment by JmillsExpensify** [comment]
Both paid out. Thank you!

**Comment by tgolen** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@rojiphil] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@rojiphil] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@rojiphil] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@rojiphil] Determine if we should create a regression test for this bug.
- [ ] [@rojiphil] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@JmillsExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by rojiphil** [comment]
* [x]  [@rojiphil] The PR that introduced the bug has been identified. Link to the PR: [Offending PR](https://github.com/Expensify/App/pull/48660)
* [x]  [@rojiphil] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: [Added comment](https://github.com/Expensify/App/pull/48660/files#r1815222500)
* [x]  [@rojiphil] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: Not Required. Existing checklist is good enough to capture such issues.
* [x]  [@rojiphil] Determine if we should create a regression test for this bug. : Yes. We can
* [x]  [@rojiphil] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again. 

Steps:
1. Go to Account Settings -> Workspaces
2. Select a workspace and go to Members->Invite new members
3. Select a member to invite and click on Next to go to the invite message view.
4. Delete the default welcome message text.
5. Verify that the message text remains empty and does not come back.

**Comment by rojiphil** [comment]
@tgolen BZ Checklist done

**Comment by rojiphil** [comment]
@tgolen The payment step is completed as mentioned [here](https://github.com/Expensify/App/issues/49899#issuecomment-2431871653). 
Can we close this issue if nothing else is pending? Thanks.

**Comment by tgolen** [comment]
Yep, we should be good here now. I'll close it out. Thanks!

**Comment by JmillsExpensify** [comment]
Test created and linked above, so closing this issue.
